lib/Parse/ParseDecl.cpp

```diff
@@ -2157,6 +2157,51 @@ bool Parser::parseTypeAttribute(TypeAttributes &Attributes, SourceLoc AtLoc,
     break;
   }
 
+  case TAK_differentiable: {
+    // Check if there is a 'linear' argument.
+    if (Tok.is(tok::l_paren) && peekToken().is(tok::identifier)) {
+      Parser::BacktrackingScope backtrack(*this);
+      consumeToken(tok::l_paren);
+
+      // Determine if we have '@differentiable(linear) (T) -> U'
+      // or '@differentiable (linear) -> U'.
+      if (Tok.getText() == "linear" && consumeIf(tok::identifier)) {
```

Comment:
These two parsing blocks seem both more complicated than necessary and less general. How about matching the pattern `(` `<identifier>` `)` followed by `(`, `@`, or `<identifier>`. That tells you the first `<identifier>` is meant to be how the type is differentiable, and you can match `linear` or produce an error.

Reply:
Thanks, that seems to simplify it a lot! Done.